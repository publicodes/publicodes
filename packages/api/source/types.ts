import type Engine from 'publicodes'
import type { PublicodesExpression } from 'publicodes'

export type Expressions = PublicodesExpression | PublicodesExpression[]

export type Situation = Partial<Record<string, PublicodesExpression>>

export type NewEngine = (
	expressions: Expressions,
	situation?: Situation
) => Engine
