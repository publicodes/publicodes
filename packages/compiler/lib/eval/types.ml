open Core
open Utils
open Utils.Output
module Any = Utils.Uid.Make ()

type concrete_type = Number | String | Bool | Date
[@@deriving sexp, compare, show]

type naked_t = Concrete of concrete_type | Any of Any.t

and t = naked_t Pos.t UnionFind.elem

let any ~pos () : t = Any (Any.mk ()) |> Pos.mk ~pos |> UnionFind.make

let concrete ~pos typ : t = Concrete typ |> Pos.mk ~pos |> UnionFind.make

let concrete_to_str = function
  | Number ->
      "un nombre"
  | String ->
      "un texte"
  | Bool ->
      "un boolÃ©en (oui / non)"
  | Date ->
      "une date"

let unify (t1 : t) (t2 : t) =
  let typ1 = t1 |> UnionFind.get in
  let typ2 = t2 |> UnionFind.get in
  match (Pos.value typ1, Pos.value typ2) with
  | Any _, Any _ ->
      return (UnionFind.union t1 t2)
  | Any _, Concrete _ ->
      return (UnionFind.merge (fun _ b -> b) t1 t2)
  | Concrete _, Any _ ->
      return (UnionFind.merge (fun a _ -> a) t1 t2)
  | Concrete c1, Concrete c2 ->
      if [%compare.equal: concrete_type] c1 c2 |> not then
        (* Todo replace with a unique type_error, with the pos of the different arguments *)
        let code, message = Err.type_incoherence in
        fatal_error ~pos:(Pos.pos typ1) ~kind:`Type ~code
          ~labels:
            [ Pos.mk ~pos:(Pos.pos typ1)
                (Format.sprintf "est %s" (concrete_to_str c1))
            ; Pos.mk ~pos:(Pos.pos typ2)
                (Format.sprintf "est %s" (concrete_to_str c2)) ]
          message
      else return t1
