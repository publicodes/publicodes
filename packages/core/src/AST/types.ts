import { ArrondiNode } from '../mecanisms/arrondi'
import { BarèmeNode } from '../mecanisms/barème'
import { ConditionNode } from '../mecanisms/condition'
import { ContextNode } from '../mecanisms/contexte'
import { DuréeNode } from '../mecanisms/durée'
import { EstNonDéfiniNode } from '../mecanisms/est'
import { EstNonApplicableNode } from '../mecanisms/est-non-applicable'
import { GrilleNode } from '../mecanisms/grille'
import { InversionNode } from '../mecanisms/inversion'
import { PossibilityNode } from '../mecanisms/one-possibility'
import { OperationNode } from '../mecanisms/operation'
import { RésoudreRéférenceCirculaireNode } from '../mecanisms/résoudre-référence-circulaire'
import { SimplifierUnitéNode } from '../mecanisms/simplifier-unité'
import { TauxProgressifNode } from '../mecanisms/tauxProgressif'
import { TexteNode } from '../mecanisms/texte'
import { UnitéNode } from '../mecanisms/unité'
import { VariableManquanteNode } from '../mecanisms/variablesManquantes'
import { VariationNode } from '../mecanisms/variations'
import { ReferenceNode } from '../reference'
import { ReplacementRule } from '../replacement'
import { RuleNode } from '../rule'

export type ConstantNode = {
	type: 'boolean' | 'number' | 'string' | undefined
	nodeValue: Evaluation
	nodeKind: 'constant'
	isNullable?: boolean
	isDefault?: boolean
}
type PossibleNodes =
	| RuleNode
	| ReferenceNode
	| ArrondiNode
	| BarèmeNode
	| DuréeNode
	| GrilleNode
	| EstNonApplicableNode
	| EstNonDéfiniNode
	| InversionNode
	| OperationNode
	| PossibilityNode
	| ContextNode
	| SimplifierUnitéNode
	| RésoudreRéférenceCirculaireNode
	| TauxProgressifNode
	| UnitéNode
	| VariationNode
	| ConditionNode
	| ConstantNode
	| ReplacementRule
	| VariableManquanteNode
	| TexteNode

export type NodeKind = PossibleNodes['nodeKind']
export type ASTNode<N extends NodeKind = NodeKind> = PossibleNodes & {
	nodeKind: N
	isDefault?: boolean
	sourceMap?: {
		mecanismName: string
		args: Record<string, ASTNode | Array<ASTNode>>
	}
	rawNode?: string | Record<string, unknown>
} & (
		| EvaluationDecoration<Types>
		// We remove the ESLINT warning as it does not concern intersection type and is actually useful here
		// https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492
		// eslint-disable-next-line @typescript-eslint/ban-types
		| {}
	)

// TODO : separate type for evaluated AST Tree

export type MecanismNode = ASTNode<
	Exclude<NodeKind, 'rule' | 'reference' | 'constant'>
>

export type ASTTransformer = (n: ASTNode) => ASTNode
export type ASTVisitor = (n: ASTNode) => void

export type TraverseFunction<K extends NodeKind> = (
	fn: ASTTransformer,
	node: ASTNode<K>,
) => ASTNode<K>

export type BaseUnit = string

// TODO: I believe it would be more effecient (for unit conversion and for
// inference), and more general to represent units using a map of base unit to
// their power number :
//
// type Unit = Map<BaseUnit, number>
// N.m²/kg² <-> {N: 1, m: 2, kg: -2} (gravity constant)
export type Unit = {
	numerators: Array<BaseUnit>
	denominators: Array<BaseUnit>
}

// Idée : une évaluation est un n-uple : (value, unit, missingVariables, isApplicable)
type EvaluationDecoration<T extends Types> = {
	nodeValue: Evaluation<T>
	unit?: Unit
	traversedVariables?: Array<string>
	missingVariables: MissingVariables
}
export type Types = number | boolean | string | Record<string, unknown>

export type Evaluation<T extends Types = Types> =
	| T
	| null // Non applicable
	| undefined // Non défini

export type EvaluatedNode<
	K extends NodeKind = NodeKind,
	T extends Types = Types,
> = EvaluationDecoration<T> & ASTNode<K>

export type MissingVariables = Record<string, number>
