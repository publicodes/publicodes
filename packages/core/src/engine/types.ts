import { Logger, PublicodesExpression } from '..'
import { type ASTNode, type EvaluatedNode } from '../AST/types'
import { getUnitKey } from '../units'

export type Cache = {
	inversionFail?: boolean
	_meta: {
		evaluationRuleStack: Array<string>
		parentRuleStack: Array<string>
		currentContexteSituation?: string
	}
	/**
	 * Every time we encounter a reference to a rule in an expression we add it
	 * to the current Set of traversed variables. Because we evaluate the
	 * expression graph “top to bottom” (ie. we start by the high-level goal and
	 * recursively evaluate its dependencies), we need to handle rule
	 * “boundaries”, so that when we “enter” in the evaluation of a dependency,
	 * we start with a clear empty set of traversed variables. Then, when we go
	 * back to the referencer rule, we need to add all to merge the two sets :
	 * rules already traversed in the current expression and the one from the
	 * reference.
	 */
	traversedVariablesStack?: Array<Set<string>>
	nodes: Map<PublicodesExpression | ASTNode, EvaluatedNode>
}

export type StrictOptions = {
	/**
	 * If set to true, the engine will throw when the
	 * situation contains invalid values
	 * (rules that don't exists, or values with syntax errors).
	 *
	 * If set to false, it will log the error and filter the invalid values from the situation
	 * @default true
	 */
	situation?: boolean

	/**
	 * If set to true, the engine will throw when parsing a rule whose parent doesn't exist
	 * This can be set to false to parse partial rule sets (e.g. optimized ones).
	 * @default true
	 */
	noOrphanRule?: boolean

	/**
	 * If set to true, the engine will throw when a cycle is detected at runtime
	 * @default false
	 */
	noCycleRuntime?: boolean

	/**
	 * If set to true, the engine will throw when a rule with 'une possibilité'
	 * is evaluated to a value that is not in the list.
	 * @default false
	 */
	checkPossibleValues?: boolean
}

export type FlagOptions = {
	/**
	 * When true, the engine will filter out possibilities that are not applicable based on the current situation.
	 * @default false
	 */
	filterNotApplicablePossibilities?: boolean
}

export type WarnOptions = {
	/**
	 * Enable warnings for experimental rules
	 * @default true
	 */
	experimentalRules?: boolean
	/**
	 * Enable warnings for unit conversion issues
	 * @default true
	 */
	unitConversion?: boolean
	/**
	 * Enable warnings for cyclic references in rules
	 * @default true
	 */
	cyclicReferences?: boolean
	/**
	 * Enable warnings for deprecated syntax
	 * @default true
	 */
	deprecatedSyntax?: boolean
	/**
	 * Enable warnings for situation issues. Warnings can be raised when the situation contains rules that are not in the base rules or when a value we try to set is not valid (it doesn't exist in the list of possible values for a given rule).
	 * @default true
	 */
	situationIssues?: boolean
}

/**
 * Options for the engine constructor
 */
export type EngineOptions = {
	/**
	 * Don't throw an error if the parent of a rule is not found.
	 * This is useful to parse partial rule sets (e.g. optimized ones).
	 * @deprecated Use the `strict: { parentRule: false }` option instead
	 */
	allowOrphanRules?: boolean
	/**
	 * Whether the engine should trigger an error when it detects an anomaly,
	 * or whether it should simply record it and continue.
	 *
	 * This option can be set globally (true or false) or for specific rules ({@link StrictOptions}).
	 *  */
	strict?: boolean | StrictOptions

	/**
	 * The logger used to log errors and warnings (default to console).
	 * @type {Logger}
	 */
	logger?: Logger

	/**
	 * getUnitKey is a function that allows to normalize the unit in the engine.
	 * @experimental
	 */
	getUnitKey?: getUnitKey

	/**
	 * flag to enable new experimental features
	 */
	flag?: FlagOptions

	/**
	 * Specify which warnings should be enabled
	 */
	warn?: boolean | WarnOptions
}
