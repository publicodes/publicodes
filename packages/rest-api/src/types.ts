import type PublicodesEngine from 'publicodes'
import type { PublicodesExpression } from 'publicodes'

export type Expressions = PublicodesExpression | PublicodesExpression[]

export type Situation = Partial<Record<string, PublicodesExpression>>

export type Engine = PublicodesEngine
