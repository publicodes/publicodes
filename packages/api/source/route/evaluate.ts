import { Expressions, NewEngine, Situation } from '../types.js'
import { catchError, PickInObject } from '../utils.js'

export interface EvaluateBody {
	expressions: Expressions
	situation?: Situation
}

export function evaluate(
	newEngine: NewEngine,
	{ expressions, situation }: EvaluateBody
) {
	const engine = newEngine(expressions, situation)
	const [error] = catchError(() => engine.setSituation(situation))

	const keysKept = [
		'nodeValue' as const,
		'unit' as const,
		'traversedVariables' as const,
		'missingVariables' as const,
	]

	const evaluateResult = (
		Array.isArray(expressions) ? expressions : [expressions]
	).map((expression) => {
		const [error, result] = catchError(() =>
			PickInObject(engine.evaluate(expression), keysKept)
		)

		return !error ? result : { error: { message: error.message } }
	})

	return {
		evaluate: evaluateResult,
		situationError: error && { message: error.message },
	}
}
