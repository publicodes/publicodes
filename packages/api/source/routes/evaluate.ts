import { Situation, NewEngine, Expressions } from '@/types'
import { catchError, PickInObject } from '@/utils'

export interface APIParameters {
	situation?: Situation
}

export function evaluate(
	newEngine: NewEngine,
	expressions: Expressions,
	{ situation }: APIParameters = {}
) {
	const engine = newEngine(expressions, situation)
	const situationResult = catchError(() => engine.setSituation(situation))

	const keysKept = [
		'nodeValue' as const,
		'unit' as const,
		'traversedVariables' as const,
		'missingVariables' as const,
	]

	const evaluateResult = (
		Array.isArray(expressions) ? expressions : [expressions]
	).map((expression) =>
		catchError(() => PickInObject(engine.evaluate(expression), keysKept))
	)

	return {
		situation: situationResult,
		evaluate: evaluateResult,
	}
}
