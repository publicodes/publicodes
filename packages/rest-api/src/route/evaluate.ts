import { Engine, Expressions, Situation } from '../types.js'
import { catchError, PickInObject } from '../utils.js'

export interface EvaluateBody {
	expressions: Expressions
	situation?: Situation
}

export function evaluate(
	originalEngine: Engine,
	{ expressions, situation }: EvaluateBody,
) {
	const engine = originalEngine.shallowCopy()
	originalEngine.subEngines = [] // This line avoid memory leak cause by multiple call to shallowCopy(), issue https://github.com/betagouv/publicodes/issues/239

	const warnings: Array<{ message: string }> = []
	if (engine.baseContext) {
		engine.baseContext.logger = {
			error: console.error,
			log: console.log,
			warn(warning: string) {
				warnings.push({ message: warning })
			},
		}
	}

	const [situationError] = catchError(() => engine.setSituation(situation))

	if (situationError) {
		return { situationError: { message: situationError.message } }
	}

	const keysKept = [
		'nodeValue' as const,
		'unit' as const,
		'traversedVariables' as const,
		'missingVariables' as const,
	]

	const evaluateResult = (
		Array.isArray(expressions) ? expressions : [expressions]).map(
		(expression) => {
			const [error, result] = catchError(() =>
				PickInObject(engine.evaluate(expression), keysKept),
			)

			return !error ? result : { error: { message: error.message }, warnings }
		},
	)

	return { evaluate: evaluateResult, warnings }
}
