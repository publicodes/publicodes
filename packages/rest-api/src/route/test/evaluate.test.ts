import Engine from 'publicodes'
import { describe, expect, it, vi } from 'vitest'
import { Expressions, Situation } from '../../types'
import { evaluate } from '../evaluate'
import { parse } from 'yaml'

const obj = {
	setSituation: (_situation?: Situation) => 42,
	evaluate: (_expressions: Expressions) => ({
		nodeValue: 42,
		unit: 42,
		traversedVariables: 42,
		missingVariables: 42,
	}),
	shallowCopy: () => obj,
}

const mockedEngine = obj as unknown as Engine

const rules = `
prix:
prix . carottes: 2€/kg
prix . champignons: 5€/kg
prix . avocat: 2€/avocat

dépenses primeur:
  formule:
    somme:
      - prix . carottes * 1.5 kg
      - prix . champignons * 500g
      - prix . avocat * 3 avocat
`

const engine = new Engine(parse(rules))

describe('evaluate', () => {
	it('Test input/output', () => {
		const spySitu = vi.spyOn(obj, 'setSituation')
		const spyEval = vi.spyOn(obj, 'evaluate')

		expect(
			evaluate(mockedEngine, {
				expressions: 'expressions',
				situation: { test: 'test' },
			}),
		).toMatchInlineSnapshot(`
			{
			  "evaluate": [
			    {
			      "missingVariables": 42,
			      "nodeValue": 42,
			      "traversedVariables": 42,
			      "unit": 42,
			    },
			  ],
			  "warnings": [],
			}
		`)

		expect(spySitu).toHaveBeenCalledWith({ test: 'test' })
		expect(spyEval).toHaveBeenCalledWith('expressions')
	})

	it('One expression in array should return same result as a single same expression', () => {
		expect(evaluate(mockedEngine, { expressions: ['coucou'] })).toMatchObject(
			evaluate(mockedEngine, { expressions: 'coucou' }),
		)
	})

	it('Simple test with real Engine', () => {
		expect(evaluate(engine, { expressions: ['21 + 21', '6 * 7'] }))
			.toMatchInlineSnapshot(`
				{
				  "evaluate": [
				    {
				      "missingVariables": {},
				      "nodeValue": 42,
				      "traversedVariables": [],
				      "unit": undefined,
				    },
				    {
				      "missingVariables": {},
				      "nodeValue": 42,
				      "traversedVariables": [],
				      "unit": undefined,
				    },
				  ],
				  "warnings": [],
				}
			`)
	})

	it('Test with real Engine', () => {
		expect(
			evaluate(engine, {
				expressions: ['dépenses primeur', 'prix', 'prix . avocat * 21'],
			}),
		).toMatchInlineSnapshot(`
			{
			  "evaluate": [
			    {
			      "missingVariables": {},
			      "nodeValue": 11.5,
			      "traversedVariables": [
			        "dépenses primeur",
			        "prix . carottes",
			        "prix . champignons",
			        "prix . avocat",
			      ],
			      "unit": {
			        "denominators": [],
			        "numerators": [
			          "€",
			        ],
			      },
			    },
			    {
			      "missingVariables": {
			        "prix": 1,
			      },
			      "nodeValue": undefined,
			      "traversedVariables": [
			        "prix",
			      ],
			      "unit": undefined,
			    },
			    {
			      "missingVariables": {},
			      "nodeValue": 42,
			      "traversedVariables": [
			        "prix . avocat",
			      ],
			      "unit": {
			        "denominators": [
			          "avocat",
			        ],
			        "numerators": [
			          "€",
			        ],
			      },
			    },
			  ],
			  "warnings": [],
			}
		`)
	})

	it('Test invalid syntax in situation', () => {
		expect(
			evaluate(engine, {
				expressions: ['1 + 1'],
				situation: { test: '"42' },
			}),
		).toMatchSnapshot()
	})

	it('Test syntax error in expression', () => {
		expect(evaluate(engine, { expressions: '1+1' })).toMatchSnapshot()
	})

	it('Test two syntax error in expression', () => {
		expect(
			evaluate(engine, {
				expressions: ['1+1', '"42', '42'],
			}),
		).toMatchSnapshot()
	})

	it('Test error in expression and situation at same time', () => {
		expect(
			evaluate(engine, { expressions: '1+1', situation: { test: '"42' } }),
		).toMatchSnapshot()
	})

	it('Test no expressions', () => {
		expect(evaluate(engine, { expressions: [] })).toMatchInlineSnapshot(`
			{
			  "evaluate": [],
			  "warnings": [],
			}
		`)
	})
})
