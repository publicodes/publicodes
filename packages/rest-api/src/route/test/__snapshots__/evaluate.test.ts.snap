// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`evaluate > Test error in expression and situation at same time 1`] = `
{
  "situationError": {
    "message": "
[ Erreur d'évaluation ]
➡️  Dans la règle \\"test\\"
✖️  Erreur lors de la mise à jour de la situation : test n'existe pas dans la base de règle.",
  },
}
`;

exports[`evaluate > Test invalid syntax in situation 1`] = `
{
  "situationError": {
    "message": "
[ Erreur d'évaluation ]
➡️  Dans la règle \\"test\\"
✖️  Erreur lors de la mise à jour de la situation : test n'existe pas dans la base de règle.",
  },
}
`;

exports[`evaluate > Test syntax error in expression 1`] = `
{
  "evaluate": [
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"$EVALUATION\\"
✖️  \`1+1\` n'est pas une expression valide
    Syntax error at line 1 col 2:

1  1+1
    ^
Unexpected additionSubstraction token: \\"+\\". Instead, I was expecting to see one of the following:

A space token based on:
    number$ebnf$1$subexpression$1 →  ● %space
    number$ebnf$1 →  ● number$ebnf$1$subexpression$1
    number → %number ● number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A word token based on:
    Words →  ● %word
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A word token based on:
    WordOrKeyword →  ● %word
    Words →  ● WordOrKeyword Words$ebnf$1
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A boolean token based on:
    WordOrKeyword →  ● %boolean
    Words →  ● WordOrKeyword Words$ebnf$1
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
A space token based on:
    UnitDenominator$ebnf$1$subexpression$1 →  ● %space
    UnitDenominator$ebnf$1 →  ● UnitDenominator$ebnf$1$subexpression$1
    UnitDenominator →  ● UnitDenominator$ebnf$1 \\"/\\" Words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"/\\" based on:
    UnitDenominator → UnitDenominator$ebnf$1 ● \\"/\\" Words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
",
      },
      "warnings": [],
    },
  ],
  "warnings": [],
}
`;

exports[`evaluate > Test two syntax error in expression 1`] = `
{
  "evaluate": [
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"$EVALUATION\\"
✖️  \`1+1\` n'est pas une expression valide
    Syntax error at line 1 col 2:

1  1+1
    ^
Unexpected additionSubstraction token: \\"+\\". Instead, I was expecting to see one of the following:

A space token based on:
    number$ebnf$1$subexpression$1 →  ● %space
    number$ebnf$1 →  ● number$ebnf$1$subexpression$1
    number → %number ● number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A word token based on:
    Words →  ● %word
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A word token based on:
    WordOrKeyword →  ● %word
    Words →  ● WordOrKeyword Words$ebnf$1
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A boolean token based on:
    WordOrKeyword →  ● %boolean
    Words →  ● WordOrKeyword Words$ebnf$1
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
A space token based on:
    UnitDenominator$ebnf$1$subexpression$1 →  ● %space
    UnitDenominator$ebnf$1 →  ● UnitDenominator$ebnf$1$subexpression$1
    UnitDenominator →  ● UnitDenominator$ebnf$1 \\"/\\" Words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"/\\" based on:
    UnitDenominator → UnitDenominator$ebnf$1 ● \\"/\\" Words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
",
      },
      "warnings": [],
    },
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"$EVALUATION\\"
✖️  \`\\"42\` n'est pas une expression valide
    invalid syntax at line 1 col 1:

1  \\"42
   ^
Unexpected input (lexer error). Instead, I was expecting to see one of the following:

A date token based on:
    Date →  ● %date
    main →  ● Date
A boolean token based on:
    NonNumericTerminal →  ● %boolean
    main →  ● NonNumericTerminal
A string token based on:
    NonNumericTerminal →  ● %string
    main →  ● NonNumericTerminal
A JSONObject token based on:
    JSONObject →  ● %JSONObject
    main →  ● JSONObject
A \\"-\\" based on:
    Negation →  ● \\"-\\" %space Parentheses
    NumericValue →  ● Negation
    main →  ● NumericValue
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" %space NumericValue %space \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" NumericValue \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A word token based on:
    Words →  ● %word
    VariableWithoutParentSelector →  ● Words VariableWithoutParentSelector$ebnf$1
    Variable →  ● VariableWithoutParentSelector
    Date →  ● Variable
    main →  ● Date
A word token based on:
    WordOrKeyword →  ● %word
    Words →  ● WordOrKeyword Words$ebnf$1
    VariableWithoutParentSelector →  ● Words VariableWithoutParentSelector$ebnf$1
    Variable →  ● VariableWithoutParentSelector
    Date →  ● Variable
    main →  ● Date
A boolean token based on:
    WordOrKeyword →  ● %boolean
    Words →  ● WordOrKeyword Words$ebnf$1
    VariableWithoutParentSelector →  ● Words VariableWithoutParentSelector$ebnf$1
    Variable →  ● VariableWithoutParentSelector
    Date →  ● Variable
    main →  ● Date
A parentSelector token based on:
    Variable$ebnf$1$subexpression$1 →  ● %parentSelector %dot
    Variable$ebnf$1 → Variable$ebnf$1 ● Variable$ebnf$1$subexpression$1
    Variable →  ● Variable$ebnf$1 VariableWithoutParentSelector
    Date →  ● Variable
    main →  ● Date
A number token based on:
    number →  ● %number
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A number token based on:
    number →  ● %number number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
",
      },
      "warnings": [],
    },
    {
      "missingVariables": {},
      "nodeValue": 42,
      "unit": undefined,
    },
  ],
  "warnings": [],
}
`;
